src/lib/datatypes/sol-buffer.c

```diff
@@ -142,8 +142,12 @@ sol_buffer_append_bytes(struct sol_buffer *buf, const uint8_t *bytes, size_t siz
     int err;
 
     SOL_NULL_CHECK(buf, -EINVAL);
-    if (size == 0)
-        return 0;
+    if (size == 0) {
+        if (SOL_BUFFER_NEEDS_NUL_BYTE(buf))
+            return sol_buffer_ensure_nul_byte(buf);
+        else
+            return 0;
+    }
```

Comment:
Not related to this patch (it was an attempt to fix the FIXME I introduce, but it's not enough). But the logic is sound -- if we get nul_bytes, any append/resize/whatever has to ensure it.